src/deepsparse/transformers/utils/decoder_kv_cache.py

```diff
@@ -103,79 +103,78 @@ def update(
-                state["state_name"]: (1, 1, 3, 1) = array([[[[3], [4], [5]]]])
-            else:
-
-                state["state_name"]: (1, 1, 3, 1) = array([[[[1], [4], [5]]]])
-            ```
-            """
-            idxs_to_remove = [
-                i + int(self._freeze_first_position)
-                for i in range(num_entries_to_delete)
-            ]
-
-            state = self._delete_entries(state, idxs_to_remove)
+        # if we had fewer padded entries than num_entries_to_delete,
+        # we additionally are forced to delete some non-padded entries (the oldest ones)
+        num_non_padded_entries_to_delete = max(
+            0, num_entries_to_delete - num_padded_entries
+        )
+
+        for name, cache_array in state.items():
+            if num_padded_entries_to_delete:
+                cache_array = self.remove_padded_entries(
+                    cache_array, num_padded_entries_to_delete
+                )
+            if num_non_padded_entries_to_delete:
+                cache_array = self.remove_non_padded_entries(
+                    cache_array, num_entries_to_delete
+                )
+            state[name] = numpy.ascontiguousarray(cache_array)
 
         self._state = state
 
+    def remove_padded_entries(
+        self, cache_array: numpy.ndarray, num_padded_entries_to_delete: int
+    ):
+        """
+        Remove the num_padded_entries_to_delete entries from the cache array.
+        This function assumes that the cache_array has the number
+        of padded (blank) entries that is equal/larger than
+        num_padded_entries_to_delete.
+
+        :param cache_array: The cache array to be modified.
+        :param num_padded_entries_to_delete: The number of padded entries to delete.
+        """
+        return cache_array[:, :, num_padded_entries_to_delete:, :]
+
+    def remove_non_padded_entries(
+        self, cache_array: numpy.ndarray, num_non_padded_entries_to_delete: int
+    ):
+        """
+        Remove the num_non_padded_entries_to_delete entries from the cache array.
+        This function assumes that the cache_array has no padded (blank) entries and
+        thus we are forced to delete the oldest entries from the cache.
+
+        If self._freeze_first_position is set to True, that means that the oldest
+        entry in the cache_array is the one that corresponds to the BOS token. Because
+        we want to keep that entry in the cache, we will delete the oldest entry
+        starting from the second oldest entry.
+        """
+        new_cache_array = cache_array[
+            :,
+            :,
+            bool(self._freeze_first_position) + num_non_padded_entries_to_delete :,
+            :,
+        ]
+        if self._freeze_first_position:
+            bos_entries = cache_array[:, :, :1, :]
+            new_cache_array = numpy.concatenate(
+                [bos_entries, new_cache_array], axis=self._sequence_len_axis
```

Comment:
could try benchmarking against idx selection

Reply:
Code used:

```python
from deepsparse.transformers.utils import DecoderKVCache
import numpy as np
import timeit

def update_session():
    session = DecoderKVCache()
    state = np.random.random((1, 32, 1000, 64))
    session.setup(session_id="dummy_id",
                  state=state,
                  num_processed_tokens=0,
                  freeze_first_position=1,
        )
    for _ in range(1000):
        session.remove_non_padded_entries(state, 500)

# run timeit on the test function
print(timeit.timeit("update_session()", setup="from __main__ import update_session", number=10))
```

Reply:
@rahul-tuli Ben probably referred to this function: https://github.com/neuralmagic/deepsparse/blob/cbc63319e3e85d2f9237559ddd0a5c7aad52735b/src/deepsparse/transformers/pipelines/text_generation.py#L422. This was how we used to (naively) do the "cache reassembly" back in the days. My code above tries to mimic this (however, avoiding the use of np.delete function). I can quickly benchmark your idea as well.

Reply:
@rahul-tuli 
Tried this implementation:
```python
B, H, L, E = cache_array.shape
new_cache_array = numpy.empty((B, H, L - num_non_padded_entries_to_delete, E))
if self._freeze_first_position:
    new_cache_array[:,:,0,:] = cache_array[:,:,0,:]
    new_cache_array[:,:,1:,:] = cache_array[:,:,num_non_padded_entries_to_delete+1:,:]
else:
    new_cache_array[:,:,:,:] = cache_array[:,:,num_non_padded_entries_to_delete:,:]
```

roughly 1,25x slower then current implementation when freezing tokens
roughly 60x slower then current implementation when not freezing tokens

Reply:
Thanks Damian! makes sense to me now :)